### PR TITLE
feat(android): Detect busy background threads

### PR DESCRIPTION
**Pull request by jonathanmos**
## Description
Espresso considers the device not idle if it has, among other things, active asynctasks, however we were not able to see these background tasks and as a result would print messages of "Device seems to be idle", instead of actually indicating the issue. In this PR, the ability to check the status of asynctasks has been added in order to provide more accurate sync debugging. 

- This pull request is associated with #3617

### CODE REVIEW

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/asynctask/DetoxBusyResource.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            val detoxBusyResourceHelper = DetoxBusyResourceHelper()
+            resourceMap.putAll(detoxBusyResourceHelper.formatResource(it))
+        }
+    }
+
+    constructor(idlingResourceDescription: IdlingResourceDescription?) {
+        idlingResourceDescription?.json()?.let {
+            resourceMap.putAll(it)
+        }
+    }
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/asynctask/DetoxBusyResource.kt:20]: Feels weird to have this just for testing. Would it make more sense to use `constructor(IdlingResource)` with a mocked idling resource?

**Review by d4vidi — CHANGES_REQUESTED** [review]
So besides the inline notes, my impression is that the roles of `DetoxBusyResourceHelper` and to some extent `DetoxBusyResource` itself have been too loosely defined. As I understand it, there are largely 2 layers involved in this mechanism - adhering to the hexagonal / port&adapter architecture:
1. The requests inbound/outbound layer, which in other words is the network "port".
2. The core, which interrogates the system's status and returns some abstract representation of the result (i.e. a list of "busy resources").

`QueryStatusActionHandler` resides in the 1st layer. The 2nd layer is `TestEngineFacade` and below. `DetoxBusyResourceHelper` seems to violate that separation by being tightly associated with `DetoxBusyResource`, on the one hand, but performing the formatting work which should stay under the jurisdiction of `QueryStatusActionHandler`.

You must reconsider this design, and provide an alternative which better aligns with the architecture.

File: detox/src/client/actions/formatters/__snapshots__/SyncStatusFormatter.test.js.snap [hunk 2]
```diff
@@ -83,13 +84,20 @@ With reasons:
...
 
 Please report this issue on our GitHub tracker:
 https://github.com/wix/Detox/issues"
 `;
 
+exports[`Sync Status Formatter busy status should format "bg" correctly 1`] = `
+"The app is busy with the following tasks:
+• Background threads are busy:
+  - Reason: asynctasks."
+`;
```
[the remark this answers is not in the document]
Comment (jonathanmos) [reply detox/src/client/actions/formatters/__snapshots__/SyncStatusFormatter.test.js.snap:99]: What if in the future we detect background work that is not caused by an asynctask? The way that this is currently written it's extensible, whereas to print a single statement locks it to asynctasks as the cause

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/asynctask/DetoxBusyResource.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        other as DetoxBusyResource
+
+        if (resourceMap != other.resourceMap) return false
+
+        return true
+    }
+
+    override fun toString(): String {
+        return "DetoxBusyResource(resourceMap=$resourceMap)"
+    }
```
[the remark this answers is not in the document]
Comment (jonathanmos) [reply detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/asynctask/DetoxBusyResource.kt:44]: This was added for unit test comparisons

**Review by d4vidi — APPROVED** [review]

**Comment by asafkorem** [comment]
Thanks @jonathanmos, Detox version [20.1.0](https://github.com/wix/Detox/releases/tag/20.1.0) has been released with this change.

**Pull request merged**